File: apps/api/src/schemas/community.py
"""
Sentinel NER — Stage 10 Community Intelligence Domain Models & Schemas
Implements:
1. CitizenReport with strict geospatial location, media validation, and lifecycle states.
2. CommunityEventCluster with explainable proximity metrics (clustering != verification).
3. CommunityModerationEvent with immutable historical audit trail.
4. Non-autonomous safety rule: reports are evidence inputs, not confirmed ground truth.
"""

from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import uuid4

from pydantic import BaseModel, Field, field_validator

from src.schemas.geojson import GeoJSONPoint


class ReportCategory(str, Enum):
    LANDSLIDE = "LANDSLIDE"
    CRACKING = "CRACKING"
    ROAD_DAMAGE = "ROAD_DAMAGE"
    DEBRIS = "DEBRIS"
    ROCKFALL = "ROCKFALL"
    DRAINAGE_BLOCKAGE = "DRAINAGE_BLOCKAGE"
    WATER_SEEPAGE = "WATER_SEEPAGE"
    SLOPE_MOVEMENT = "SLOPE_MOVEMENT"
    OTHER = "OTHER"


class ReportStatus(str, Enum):
    SUBMITTED = "SUBMITTED"
    PROCESSING = "PROCESSING"
    UNVERIFIED = "UNVERIFIED"
    PROBABLE = "PROBABLE"
    VERIFIED = "VERIFIED"
    ACTIONED = "ACTIONED"
    RESOLVED = "RESOLVED"
    REJECTED = "REJECTED"


class ModerationState(str, Enum):
    UNMODERATED = "UNMODERATED"
    IN_REVIEW = "IN_REVIEW"
    MODERATED = "MODERATED"


class ReportSource(str, Enum):
    MOBILE_APP = "MOBILE_APP"
    WEB_PORTAL = "WEB_PORTAL"
    FIELD_TERMINAL = "FIELD_TERMINAL"
    HOTLINE = "HOTLINE"
    OFFLINE_SYNC = "OFFLINE_SYNC"


class LocationSource(str, Enum):
    DEVICE_GPS = "DEVICE_GPS"
    MANUAL_PIN = "MANUAL_PIN"
    MAP_SELECTION = "MAP_SELECTION"
    FIELD_SURVEY = "FIELD_SURVEY"


class MediaReference(BaseModel):
    """
    Validated object-storage reference for community images/media.
    Never stores large binary blobs directly in MongoDB.
    """
    object_key: str = Field(..., description="S3/Blob storage key")
    content_type: str = Field(..., description="MIME type, e.g. image/jpeg, image/png")
    size_bytes: int = Field(..., ge=1, le=25_000_000, description="Payload size limit 25MB")
    checksum_sha256: str = Field(..., min_length=64, max_length=64, description="SHA-256 integrity digest")
    uploaded_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    provenance: Dict[str, Any] = Field(default_factory=dict)
    is_verified_safe: bool = Field(default=True, description="Passed virus and polyglot scan")

    @field_validator("content_type")
    @classmethod
    def validate_mime(cls, v: str) -> str:
        allowed = {"image/jpeg", "image/png", "image/webp", "image/heic", "video/mp4"}
        if v.lower() not in allowed:
            raise ValueError(f"Content type '{v}' not permitted. Must be one of {sorted(allowed)}")
        return v.lower()

    @field_validator("object_key")
    @classmethod
    def validate_key(cls, v: str) -> str:
        if ".." in v or v.startswith("/") or "\\" in v:
            raise ValueError("Path traversal sequences prohibited in media object key")
        return v


class CitizenReport(BaseModel):
    """
    Authoritative Citizen Observation Record.
    Evidence input into the intelligence system; does not constitute confirmed fact without review.
    """
    id: str = Field(default_factory=lambda: f"cr-{uuid4().hex[:12]}")
    reporter_id: str = Field(..., description="User ID or anonymous contributor hash")
    organization_id: Optional[str] = Field(None, description="Optional tenant or agency")
    district_id: str = Field(..., description="Authoritative district identifier (e.g. AIZAWL)")
    location: GeoJSONPoint = Field(..., description="Point location EPSG:4326")
    location_accuracy_m: Optional[float] = Field(None, ge=0.0, description="Accuracy radius in meters")
    location_source: LocationSource = Field(default=LocationSource.DEVICE_GPS)
    coordinate_reference: str = Field(default="EPSG:4326")
    reported_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    received_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    category: ReportCategory = Field(...)
    description: Optional[str] = Field(None, max_length=2000)
    media_references: List[MediaReference] = Field(default_factory=list, max_length=5)
    source: ReportSource = Field(default=ReportSource.WEB_PORTAL)
    status: ReportStatus = Field(default=ReportStatus.SUBMITTED)
    confidence: float = Field(default=0.5, ge=0.0, le=1.0)
    moderation_state: ModerationState = Field(default=ModerationState.UNMODERATED)
    reviewer_id: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    rejection_reason: Optional[str] = None
    linked_entities: Dict[str, Any] = Field(
        default_factory=dict,
        description="Associated slope_unit_id, road_id, chainage_km, village_id, asset_id"
    )
    provenance: Dict[str, Any] = Field(default_factory=dict)
    correlation_id: str = Field(default_factory=lambda: f"corr-{uuid4().hex[:10]}")


class CitizenReportCreate(BaseModel):
    """Input payload for submitting a citizen hazard report."""
    district_id: str
    location: GeoJSONPoint
    location_accuracy_m: Optional[float] = None
    location_source: LocationSource = LocationSource.DEVICE_GPS
    category: ReportCategory
    description: Optional[str] = Field(None, max_length=2000)
    media_references: List[MediaReference] = Field(default_factory=list, max_length=5)
    source: ReportSource = ReportSource.WEB_PORTAL
    linked_entities: Optional[Dict[str, Any]] = None


class CommunityModerationEvent(BaseModel):
    """
    Immutable audit trail entry for citizen report moderation.
    Every status change (VERIFIED, PROBABLE, REJECTED) must be documented here.
    """
    id: str = Field(default_factory=lambda: f"mod-{uuid4().hex[:12]}")
    report_id: str
    moderator_id: str
    moderator_name: str
    old_status: ReportStatus
    new_status: ReportStatus
    reason: str = Field(..., min_length=5, max_length=1000)
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    evidence_references: List[str] = Field(default_factory=list)
    correlation_id: str = Field(default_factory=lambda: f"corr-{uuid4().hex[:10]}")


class ReportModerationRequest(BaseModel):
    new_status: ReportStatus = Field(..., description="Target status: VERIFIED, PROBABLE, REJECTED, RESOLVED")
    reason: str = Field(..., min_length=5, max_length=1000)
    evidence_references: List[str] = Field(default_factory=list)


class CommunityEventCluster(BaseModel):
    """
    Explainable group of spatially and temporally correlated citizen reports.
    RULE: A cluster indicates elevated observation density requiring review, NEVER automatic confirmation.
    """
    id: str = Field(default_factory=lambda: f"cls-{uuid4().hex[:12]}")
    cluster_code: str
    district_id: str
    center_point: GeoJSONPoint
    radius_meters: float
    category: ReportCategory
    report_ids: List[str]
    report_count: int
    confidence_score: float = Field(..., ge=0.0, le=1.0)
    explanation: str
    first_reported_at: datetime
    last_reported_at: datetime
    status: str = Field(default="REQUIRES_REVIEW", description="REQUIRES_REVIEW, REVIEWED, RESOLVED")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class CommunitySummary(BaseModel):
    total_reports: int
    submitted_count: int
    unverified_count: int
    probable_count: int
    verified_count: int
    rejected_count: int
    resolved_count: int
    cluster_count: int
    district_id: Optional[str] = None
    generated_at: datetime

File: apps/web/src/lib/community.ts
/**
 * Sentinel NER — Stage 10 Community Intelligence Typed API Client
 */

import { fetchFromAPI } from './api';

export type ReportCategory =
  | 'LANDSLIDE'
  | 'CRACKING'
  | 'ROAD_DAMAGE'
  | 'DEBRIS'
  | 'ROCKFALL'
  | 'DRAINAGE_BLOCKAGE'
  | 'WATER_SEEPAGE'
  | 'SLOPE_MOVEMENT'
  | 'OTHER';

export type ReportStatus =
  | 'SUBMITTED'
  | 'PROCESSING'
  | 'UNVERIFIED'
  | 'PROBABLE'
  | 'VERIFIED'
  | 'ACTIONED'
  | 'RESOLVED'
  | 'REJECTED';

export type ModerationState = 'UNMODERATED' | 'IN_REVIEW' | 'MODERATED';

export interface MediaReference {
  object_key: string;
  content_type: string;
  size_bytes: number;
  checksum_sha256: string;
  uploaded_at: string;
  provenance?: Record<string, any>;
  is_verified_safe?: boolean;
}

export interface CitizenReport {
  id: string;
  reporter_id: string;
  organization_id?: string | null;
  district_id: string;
  location: {
    type: string;
    coordinates: [number, number]; // [lng, lat]
  };
  location_accuracy_m?: number | null;
  location_source: string;
  coordinate_reference: string;
  reported_at: string;
  received_at: string;
  category: ReportCategory;
  description?: string | null;
  media_references: MediaReference[];
  source: string;
  status: ReportStatus;
  confidence: number;
  moderation_state: ModerationState;
  reviewer_id?: string | null;
  reviewed_at?: string | null;
  rejection_reason?: string | null;
  linked_entities?: Record<string, any>;
  provenance?: Record<string, any>;
  correlation_id: string;
  moderation_history?: Array<{
    id: string;
    moderator_id: string;
    moderator_name: string;
    old_status: ReportStatus;
    new_status: ReportStatus;
    reason: string;
    timestamp: string;
  }>;
}

export interface CommunityEventCluster {
  id: string;
  cluster_code: string;
  district_id: string;
  center_point: {
    type: string;
    coordinates: [number, number];
  };
  radius_meters: number;
  category: ReportCategory;
  report_ids: string[];
  report_count: number;
  confidence_score: number;
  explanation: string;
  first_reported_at: string;
  last_reported_at: string;
  status: string;
  created_at: string;
}

export interface CommunitySummary {
  total_reports: number;
  submitted_count: number;
  unverified_count: number;
  probable_count: number;
  verified_count: number;
  rejected_count: number;
  resolved_count: number;
  cluster_count: number;
  district_id?: string | null;
  generated_at: string;
}

export interface CitizenReportCreate {
  district_id: string;
  location: {
    type: string;
    coordinates: [number, number];
  };
  location_accuracy_m?: number;
  location_source?: string;
  category: ReportCategory;
  description?: string;
  media_references?: MediaReference[];
  source?: string;
  linked_entities?: Record<string, any>;
}

export async function fetchCitizenReports(params: {
  district_id?: string;
  category?: string;
  status?: string;
  moderation_state?: string;
  skip?: number;
  limit?: number;
} = {}): Promise<{ items: CitizenReport[]; total: number; skip: number; limit: number }> {
  const q = new URLSearchParams();
  if (params.district_id) q.set('district_id', params.district_id);
  if (params.category) q.set('category', params.category);
  if (params.status) q.set('status', params.status);
  if (params.moderation_state) q.set('moderation_state', params.moderation_state);
  if (params.skip != null) q.set('skip', String(params.skip));
  if (params.limit != null) q.set('limit', String(params.limit));

  const res = await fetchFromAPI<{ success: boolean; data: { items: CitizenReport[]; total: number; skip: number; limit: number } }>(
    `/api/v1/community/reports?${q.toString()}`
  );
  return res.data;
}

export async function fetchCommunitySummary(districtId?: string): Promise<CommunitySummary> {
  const q = districtId ? `?district_id=${districtId}` : '';
  const res = await fetchFromAPI<{ success: boolean; data: CommunitySummary }>(
    `/api/v1/community/reports/summary${q}`
  );
  return res.data;
}

export async function fetchCitizenReport(reportId: string): Promise<CitizenReport> {
  const res = await fetchFromAPI<{ success: boolean; data: CitizenReport }>(
    `/api/v1/community/reports/${reportId}`
  );
  return res.data;
}

export async function submitCitizenReport(payload: CitizenReportCreate): Promise<CitizenReport> {
  const res = await fetchFromAPI<{ success: boolean; data: CitizenReport; message: string }>(
    '/api/v1/community/reports',
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    }
  );
  return res.data;
}

export async function moderateCitizenReport(
  reportId: string,
  newStatus: ReportStatus,
  reason: string,
  evidenceReferences: string[] = []
): Promise<CitizenReport> {
  const res = await fetchFromAPI<{ success: boolean; data: CitizenReport; message: string }>(
    `/api/v1/community/reports/${reportId}/moderate`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        new_status: newStatus,
        reason,
        evidence_references: evidenceReferences,
      }),
    }
  );
  return res.data;
}

export async function fetchCommunityClusters(districtId?: string): Promise<CommunityEventCluster[]> {
  const q = districtId ? `?district_id=${districtId}` : '';
  const res = await fetchFromAPI<{ success: boolean; data: CommunityEventCluster[] }>(
    `/api/v1/community/clusters${q}`
  );
  return res.data;
}

export async function recalculateCommunityClusters(districtId: string): Promise<{
  district_id: string;
  clusters_generated: number;
  clusters: CommunityEventCluster[];
}> {
  const res = await fetchFromAPI<{
    success: boolean;
    data: { district_id: string; clusters_generated: number; clusters: CommunityEventCluster[] };
    message: string;
  }>(`/api/v1/community/clusters/recalculate?district_id=${districtId}`, {
    method: 'POST',
  });
  return res.data;
}

/* ═══════════════════════════════════════════════════════════════════════════
   STAGE 10: GEOFENCED PRONE AREA PROXIMITY & SMS ALERT SERVICE
   ═══════════════════════════════════════════════════════════════════════════ */

export interface SMSAlertEnrollment {
  phoneNumber: string;
  recipientName?: string;
  preferredLanguage: 'en' | 'hi' | 'mizo';
  districtId?: string;
  isActive: boolean;
  enrolledAt: string;
  lastNotifiedAt?: string;
}

export interface ProneAreaZone {
  id: string;
  name: string;
  corridorName: string;
  districtId: string;
  state: string;
  center: [number, number]; // [lat, lng]
  dangerRadiusMeters: number; // e.g. 500m (Red Zone)
  warningRadiusMeters: number; // e.g. 2500m (Advisory Zone)
  riskLevel: 'CRITICAL' | 'HIGH' | 'MODERATE';
  hazardType: ReportCategory;
  currentSituation: string; // What is currently happening in the affected prone area
  roadStatus: string;
  recommendedAction: string;
  activeSince: string;
}

export interface SMSAlertMessage {
  id: string;
  proneAreaId: string;
  proneAreaName: string;
  recipientPhone: string;
  senderId: string;
  messageText: string;
  riskLevel: 'CRITICAL' | 'HIGH' | 'MODERATE';
  distanceMeters: number;
  dispatchedAt: string;
  acknowledged?: boolean;
}

/** Authoritative Northeast Landslide Prone Areas & Corridors */
export const PRONE_AREAS_NER: ProneAreaZone[] = [
  {
    id: 'zone-durtlang-01',
    name: 'Durtlang Ridge Corridor',
    corridorName: 'NH-54 km 44.2 (North Bend)',
    districtId: 'dst-aizawl',
    state: 'Mizoram',
    center: [23.7548, 92.7214],
    dangerRadiusMeters: 600,
    warningRadiusMeters: 2500,
    riskLevel: 'HIGH',
    hazardType: 'LANDSLIDE',
    currentSituation: 'Active scarp slip & progressive clay detachment. Approx 15cm road surface displacement with continuous rubble fall across northbound lane.',
    roadStatus: 'SINGLE LANE CONVOY ONLY • HEAVY TRUCKS DIVERTED',
    recommendedAction: 'Reduce speed to 15 km/h. Keep clear of cliff toe. Follow PWD / BRO Pushpak flaggers or divert via Sairang.',
    activeSince: '2026-09-08T06:30:00Z',
  },
  {
    id: 'zone-ranipool-02',
    name: 'Ranipool – Singtam Corridor',
    corridorName: 'NH-10 Himalayan Sector',
    districtId: 'dst-east-sikkim',
    state: 'Sikkim',
    center: [27.3389, 88.6138],
    dangerRadiusMeters: 750,
    warningRadiusMeters: 3000,
    riskLevel: 'CRITICAL',
    hazardType: 'ROCKFALL',
    currentSituation: 'Recurring boulder detachments from vertical gneissic cut slope above highway. High rainfall triggering rapid mudwash.',
    roadStatus: 'INTERMITTENT BLOCKAGE • ESCORTED TRANSIT',
    recommendedAction: 'Do not stop vehicles in rockfall chute zone. Evacuate roadside huts. Emergency machinery deployed by BRO Swastik.',
    activeSince: '2026-09-08T11:15:00Z',
  },
  {
    id: 'zone-ramhlun-03',
    name: 'Ramhlun North Tension Ridge',
    corridorName: 'Ramhlun Residential Spur',
    districtId: 'dst-aizawl',
    state: 'Mizoram',
    center: [23.7489, 92.7301],
    dangerRadiusMeters: 450,
    warningRadiusMeters: 2000,
    riskLevel: 'MODERATE',
    hazardType: 'CRACKING',
    currentSituation: 'Longitudinal tension cracks expanding across residential retaining wall and roadway following 65mm precipitation.',
    roadStatus: 'CAUTION ADVISORY • LIGHT VEHICLES ONLY',
    recommendedAction: 'Monitor building foundation cracks. Divert heavy axle loads. Report widening to Disaster Volunteer network.',
    activeSince: '2026-09-09T03:00:00Z',
  },
  {
    id: 'zone-bilkhawthlir-04',
    name: 'Bilkhawthlir Lowland Corridor',
    corridorName: 'NH-306 Valley Spur',
    districtId: 'dst-kolasib',
    state: 'Mizoram',
    center: [24.1628, 92.6845],
    dangerRadiusMeters: 500,
    warningRadiusMeters: 2200,
    riskLevel: 'HIGH',
    hazardType: 'WATER_SEEPAGE',
    currentSituation: 'Sudden high-volume turbid spring emergence at slope toe with active soil liquefaction and drainage ditch overflow.',
    roadStatus: 'FLOODED SLOPES • WATERLOGGED EMBANKMENT',
    recommendedAction: 'Avoid driving through muddy slope runoff. Watch for sudden road drop. PWD drainage team on site.',
    activeSince: '2026-09-08T18:45:00Z',
  },
  {
    id: 'zone-zote-05',
    name: 'Zote Access Corridor',
    corridorName: 'Champhai Bypass km 12',
    districtId: 'dst-champhai',
    state: 'Mizoram',
    center: [23.4721, 93.3289],
    dangerRadiusMeters: 400,
    warningRadiusMeters: 2000,
    riskLevel: 'MODERATE',
    hazardType: 'ROAD_DAMAGE',
    currentSituation: 'Step-settlement of 18cm along road foundation with shear slip cracks extending into agricultural terraces.',
    roadStatus: 'SLOW TRANSIT (MAX 10 KM/H)',
    recommendedAction: 'Follow flaggers. Avoid parking on outer embankment shoulder.',
    activeSince: '2026-09-09T05:20:00Z',
  },
];

/**
 * Calculates geodesic distance between two [latitude, longitude] coordinates using Haversine formula.
 * Returns distance in meters.
 */
export function calculateDistanceMeters(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const R = 6371000; // Earth radius in meters
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return Math.round(R * c);
}

/**
 * Formats standard Government Common Alerting Protocol (CAP) emergency SMS payload.
 */
export function generateProneAreaSMS(
  zone: ProneAreaZone,
  phone: string,
  distanceMeters: number
): SMSAlertMessage {
  const isInsideDanger = distanceMeters <= zone.dangerRadiusMeters;
  const distText =
    distanceMeters < 1000 ? `${distanceMeters}m` : `${(distanceMeters / 1000).toFixed(1)}km`;

  const urgencyTag = isInsideDanger
    ? 'CRITICAL RED ALERT • IMMEDIATE DANGER'
    : 'PROXIMITY WARNING • HAZARD ZONE APPROACH';

  const messageText = `[GOI-NDMA / NLEWS EMERGENCY SMS ALERT]
URGENCY: ${urgencyTag}
LOCATION: ${zone.name} (${zone.corridorName}, ${zone.state})
PROXIMITY: You are currently ${distText} from the affected active landslide prone zone.
WHAT IS HAPPENING: ${zone.currentSituation}
ROAD STATUS: ${zone.roadStatus}
ACTION REQUIRED: ${zone.recommendedAction}
EMERGENCY HELPLINE: 1078 (NDMA Toll-Free 24/7) • Ref: CAP-NER-${zone.id.toUpperCase()}-${Date.now().toString().slice(-4)}`;

  return {
    id: `sms-${Date.now()}-${Math.random().toString(36).slice(2, 6)}`,
    proneAreaId: zone.id,
    proneAreaName: zone.name,
    recipientPhone: phone,
    senderId: 'GOI-NDMA',
    messageText,
    riskLevel: zone.riskLevel,
    distanceMeters,
    dispatchedAt: new Date().toISOString(),
    acknowledged: false,
  };
}
